python/ray/serve/tests/unit/test_singleton_thread_router.py

```diff
@@ -0,0 +1,266 @@
+
+    mock_router = Mock(spec=AsyncioRouter)
+
+    async def mock_assign_request(*args, **kwargs):
+        return mock_replica_result
+
+    mock_router.assign_request.side_effect = mock_assign_request
+
+    with patch("ray.serve._private.router.AsyncioRouter", return_value=mock_router):
+        router = SingletonThreadRouter(
+            controller_handle=Mock(),
+            deployment_id=DeploymentID(name="test", app_name="test"),
+            handle_id="test",
+            self_actor_id="test",
+            handle_source="test",
+            replica_scheduler=Mock(),
+            enable_strict_max_ongoing_requests=False,
+            resolve_request_arg_func=Mock(),
+        )
+
+        request_meta = RequestMetadata(
+            request_id="test",
+            internal_request_id="test",
+        )
+
+        future = router.assign_request(request_meta)
+        assert isinstance(future, concurrent.futures.Future)
+        assert future.result() == mock_replica_result
+
+
+@pytest.mark.asyncio
+async def test_cancellation_propagation():
+    mock_replica_result = Mock(spec=ReplicaResult)
+    mock_replica_result.cancel = Mock()
+
+    loop = SingletonThreadRouter._get_singleton_asyncio_loop()
+
+    async def init_events():
+        return asyncio.Event(), asyncio.Event()
+
+    f = asyncio.run_coroutine_threadsafe(init_events(), loop)
+    lock_acquire_event, cancel_block_event = f.result()
+
+    mock_router = Mock(spec=AsyncioRouter)
+
+    async def mock_assign_request(*args, **kwargs):
+        try:
+            await lock_acquire_event.wait()
+            return mock_replica_result
+        except asyncio.CancelledError:
+            cancel_block_event.set()
+            raise
+
+    mock_router.assign_request.side_effect = mock_assign_request
+
+    with patch("ray.serve._private.router.AsyncioRouter", return_value=mock_router):
+        router = SingletonThreadRouter(
+            controller_handle=Mock(),
+            deployment_id=DeploymentID(name="test", app_name="test"),
+            handle_id="test",
+            self_actor_id="test",
+            handle_source="test",
+            replica_scheduler=Mock(),
+            enable_strict_max_ongoing_requests=False,
+        )
+
+        request_meta = RequestMetadata(
+            request_id="test",
+            internal_request_id="test",
+        )
+
+        assign_request_future = router.assign_request(request_meta)
+        assign_request_future.cancel()
+
+        async def coro():
+            await cancel_block_event.wait()
+
+        asyncio.run_coroutine_threadsafe(coro(), loop).result()
+
+        def release_lock():
+            lock_acquire_event.set()
+
+        loop.call_soon_threadsafe(release_lock)
+
+        with pytest.raises(concurrent.futures.CancelledError):
+            assign_request_future.result()
+
+        # ensure call count remains 0 after 1 second
+        await asyncio.sleep(1)
+        assert mock_replica_result.cancel.call_count == 0
```

Comment:
is this necessary? sleeps like this make the unit tests slow which degrades developer productivity

checking call counts and other internal implementation details is a bit of an antipattern regardless

Reply:
Sleep was not necessary, I was able to remove it and make the test more robust.

I don't share the same opinion on `checking call counts and other internal implementation details is a bit of an antipattern`, so keeping that here. We should talk about it in person. :)